### ISSUE TITLE
Latest coverities

### ISSUE THREAD

**Issue by radare** [radare/radare2#4862]
```

Hi,

Please find the latest report on new defect(s) introduced to radare2 found with Coverity Scan.

11 new defect(s) introduced to radare2 found with Coverity Scan.
4 defect(s), reported by Coverity Scan earlier, were marked fixed in the recent build analyzed by Coverity Scan.

New defect(s) Reported-by: Coverity Scan
Showing 11 of 11 defect(s)


** CID 1355196:  Integer handling issues  (CONSTANT_EXPRESSION_RESULT)
/libr/anal/p/anal_arc.c: 221 in arcompact_genops()


________________________________________________________________________________________________________
*** CID 1355196:  Integer handling issues  (CONSTANT_EXPRESSION_RESULT)
/libr/anal/p/anal_arc.c: 221 in arcompact_genops()
215     		if (fields.format == 2) {
216     			op->type = R_ANAL_OP_TYPE_MOV;
217     			op->val = sex_s12 (fields.a << 6 | fields.c);
218     		} else if (fields.format == 3) {
219     			op->type = R_ANAL_OP_TYPE_CMOV;
220     			/* TODO: cond codes */
   CID 1355196:  Integer handling issues  (CONSTANT_EXPRESSION_RESULT)
   "(fields.a & 0x20) == 1" is always false regardless of the values of its operands. This occurs as the logical operand of if.
221     			if ((fields.a & 0x20) == 1) {
222     				/* its a move from imm u6 */
223     				op->val = fields.c;
224     			} else if (fields.c == ARC_REG_LIMM) {
225     				/* its a move from limm */
226     				op->val = fields.limm;

** CID 1355206:  Incorrect expression  (SIZEOF_MISMATCH)
/libr/bin/format/mach0/mach0.c: 401 in parse_signature()


________________________________________________________________________________________________________
*** CID 1355206:  Incorrect expression  (SIZEOF_MISMATCH)
/libr/bin/format/mach0/mach0.c: 401 in parse_signature()
395     	data = link.dataoff;
396     	if (data > bin->size || data + sizeof(struct super_blob_t) > bin->size)
397     	    	return;
398     	struct super_blob_t *super = (struct super_blob_t *) (bin->b->buf + data);
399     	count = r_read_ble32 (&super->count, true);
400     	for (index = 0; index < count; ++index) {
   CID 1355206:  Incorrect expression  (SIZEOF_MISMATCH)
   Adding "index * 8UL /* sizeof (struct blob_index_t) */" to pointer "super->index" of type "struct blob_index_t []" is suspicious because adding an integral value to this pointer automatically scales that value by the size, 8 bytes, of the pointed-to type, "struct blob_index_t".  Most likely, the multiplication by "sizeof (struct blob_index_t)" in this expression is extraneous and should be eliminated.
401     		if ((ut8 *)(super->index +
402     			    index * sizeof (struct blob_index_t)) >
403     		    (ut8 *)(bin->b->buf + bin->size))
404     			return;
405     		if (r_read_ble32 (&super->index[index].type, true) == CSSLOT_ENTITLEMENTS) {
406     			ut32 begin = r_read_ble32 (&super->index[index].offset, true);

** CID 1355205:    (SIGN_EXTENSION)
/libr/anal/p/anal_arc.c: 428 in arcompact_op()
/libr/anal/p/anal_arc.c: 473 in arcompact_op()


________________________________________________________________________________________________________
*** CID 1355205:    (SIGN_EXTENSION)
/libr/anal/p/anal_arc.c: 428 in arcompact_op()
422     			/* Branch Conditionally 0x00 [0x0] */
423     			fields.limm = sex_s21 (fields.limm);
424     			op->type = R_ANAL_OP_TYPE_CJMP;
425     			/* TODO: cond codes */
426     		} else {
427     			/* Branch Unconditional Far 0x00 [0x1] */
   CID 1355205:    (SIGN_EXTENSION)
   Suspicious implicit sign extension: "fields.c" with type "unsigned short" (16 bits, unsigned) is promoted in "fields.c << 21" to type "int" (32 bits, signed), then sign-extended to type "long long" (64 bits, signed).  If "fields.c << 21" is greater than 0x7FFFFFFF, the upper bits of the result will all be 1.
428     			fields.limm |= fields.c << 21;
429     			fields.limm = sex_s25 (fields.limm);
430     			op->type = R_ANAL_OP_TYPE_JMP;
431     		}
432     		op->jump = (addr & ~3) + fields.limm;
433     		op->fail = addr + op->size;
/libr/anal/p/anal_arc.c: 473 in arcompact_op()
467     				/* Branch and Link Conditionally, 0x01, [0x0, 0x0] */
468     				fields.limm = sex_s21 (fields.limm);
469     				op->type = R_ANAL_OP_TYPE_CCALL;
470     				/* TODO: cond codes */
471     			} else {
472     				/* Branch and Link Unconditional Far, 0x01, [0x0, 0x1] */
   CID 1355205:    (SIGN_EXTENSION)
   Suspicious implicit sign extension: "fields.c" with type "unsigned short" (16 bits, unsigned) is promoted in "fields.c << 21" to type "int" (32 bits, signed), then sign-extended to type "long long" (64 bits, signed).  If "fields.c << 21" is greater than 0x7FFFFFFF, the upper bits of the result will all be 1.
473     				fields.limm |= fields.c << 21;
474     				fields.limm = sex_s25 (fields.limm);
475     				op->type = R_ANAL_OP_TYPE_CALL;
476     			}
477     			op->jump = (addr & ~3) + fields.limm;
478     			op->fail = addr + op->size;

** CID 1355204:  Resource leaks  (RESOURCE_LEAK)
/libr/util/punycode.c: 117 in r_punycode_encode()


________________________________________________________________________________________________________
*** CID 1355204:  Resource leaks  (RESOURCE_LEAK)
/libr/util/punycode.c: 117 in r_punycode_encode()
111     			if ((ut8)src[si] >= n && (ut8)src[si] <= m) {
112     				m = src[si];
113     			}
114     		}
115     
116     		if ((unsigned int)(m - n) > (UT32_MAX - delta) / (h + 1)) {
   CID 1355204:  Resource leaks  (RESOURCE_LEAK)
   Variable "dst" going out of scope leaks the storage it points to.
117     			return NULL;
118     		}
119     
120     		delta += (m - n) * (h + 1);
121     		n = m;
122     

** CID 1355203:  Resource leaks  (RESOURCE_LEAK)
/libr/util/punycode.c: 160 in r_punycode_decode()


________________________________________________________________________________________________________
*** CID 1355203:  Resource leaks  (RESOURCE_LEAK)
/libr/util/punycode.c: 160 in r_punycode_decode()
154     	if (!dst) {
155     		return NULL;
156     	}
157     
158     	for (si = 0; si < srclen; si++) {
159     		if (src[si] & 0x80) {
   CID 1355203:  Resource leaks  (RESOURCE_LEAK)
   Variable "dst" going out of scope leaks the storage it points to.
160     			return NULL;       /*just return it*/
161     		}
162     	}
163     
164     	for (p = src + srclen - 1; p > src && *p != '-'; p--);
165     	b = p - src;

** CID 1355202:  Control flow issues  (NO_EFFECT)
/libr/anal/esil.c: 1912 in esil_mem_modeq_n()


________________________________________________________________________________________________________
*** CID 1355202:  Control flow issues  (NO_EFFECT)
/libr/anal/esil.c: 1912 in esil_mem_modeq_n()
1906     		} else {
1907     			r_anal_esil_push (esil, dst);
1908     			ret = (!!esil_peek_n (esil, bits));
1909     			src1 = r_anal_esil_pop (esil);
1910     			if (src1 && r_anal_esil_get_parm (esil, src1, &d)) {
1911     				ut64 mod = d % s;
   CID 1355202:  Control flow issues  (NO_EFFECT)
   This less-than-zero comparison of an unsigned value is never true. "d < 0ULL".
1912     				mod += d < 0? s: 0;
1913     				r_anal_esil_pushnum (esil, mod);
1914     				r_anal_esil_push (esil, dst);
1915     				ret &= (!!esil_poke_n (esil, bits));
1916     			} else ret = 0;
1917     		}

** CID 1355201:  Control flow issues  (NO_EFFECT)
/libr/anal/esil.c: 1137 in esil_modeq()


________________________________________________________________________________________________________
*** CID 1355201:  Control flow issues  (NO_EFFECT)
/libr/anal/esil.c: 1137 in esil_modeq()
1131     					esil->old = d;
1132     					esil->cur = d % s;
1133     					esil->cur += d < 0? s: 0;
1134     					esil->lastsz = esil_internal_sizeof_reg (esil, dst);
1135     				}
1136     				ut64 mod = d % s;
   CID 1355201:  Control flow issues  (NO_EFFECT)
   This less-than-zero comparison of an unsigned value is never true. "d < 0ULL".
1137     				mod += d < 0? s: 0;
1138     				r_anal_esil_reg_write (esil, dst, mod);
1139     			} else {
1140     				ERR ("esil_modeq: Division by zero!");
1141     				esil->trap = R_ANAL_TRAP_DIVBYZERO;
1142     				esil->trap_code = 0;

** CID 1355200:  Control flow issues  (NO_EFFECT)
/libr/anal/esil.c: 1109 in esil_mod()


________________________________________________________________________________________________________
*** CID 1355200:  Control flow issues  (NO_EFFECT)
/libr/anal/esil.c: 1109 in esil_mod()
1103     			if (s == 0) {
1104     				eprintf ("0x%08"PFMT64x" esil_mod: Division by zero!\n", esil->address);
1105     				esil->trap = R_ANAL_TRAP_DIVBYZERO;
1106     				esil->trap_code = 0;
1107     			} else {
1108     				ut64 mod = d % s;
   CID 1355200:  Control flow issues  (NO_EFFECT)
   This less-than-zero comparison of an unsigned value is never true. "d < 0ULL".
1109     				mod += d < 0? s: 0;
1110     				r_anal_esil_pushnum (esil, mod);
1111     			}
1112     			ret = 1;
1113     		}
1114     	} else {

** CID 1355199:  Program hangs  (INFINITE_LOOP)
/libr/util/punycode.c: 18 in adapt_bias()


________________________________________________________________________________________________________
*** CID 1355199:  Program hangs  (INFINITE_LOOP)
/libr/util/punycode.c: 18 in adapt_bias()
12     
13     ut32 adapt_bias(ut32 delta, unsigned n_points, int is_first) {
14     	ut32 k;
15     	delta /= is_first ? DAMP : 2;
16     	delta += delta / n_points;
17     
   CID 1355199:  Program hangs  (INFINITE_LOOP)
   If "k > 455U" is initially true then it will remain true.
18     	for (k = 0; k > ((BASE - TMIN) * TMAX) / 2; k += BASE) {
19     		delta /= (BASE - TMIN);
20     	}
21     
22     	return k + (((BASE - TMIN + 1) * delta) / (delta + SKEW));
23     }

** CID 1355198:  Control flow issues  (DEADCODE)
/libr/util/punycode.c: 19 in adapt_bias()


________________________________________________________________________________________________________
*** CID 1355198:  Control flow issues  (DEADCODE)
/libr/util/punycode.c: 19 in adapt_bias()
13     ut32 adapt_bias(ut32 delta, unsigned n_points, int is_first) {
14     	ut32 k;
15     	delta /= is_first ? DAMP : 2;
16     	delta += delta / n_points;
17     
18     	for (k = 0; k > ((BASE - TMIN) * TMAX) / 2; k += BASE) {
   CID 1355198:  Control flow issues  (DEADCODE)
   Execution cannot reach this statement: "delta /= 35U;".
19     		delta /= (BASE - TMIN);
20     	}
21     
22     	return k + (((BASE - TMIN + 1) * delta) / (delta + SKEW));
23     }
24     

** CID 1355197:  Control flow issues  (DEADCODE)
/libr/crypto/p/crypto_rc6.c: 33 in rc6_init()


________________________________________________________________________________________________________
*** CID 1355197:  Control flow issues  (DEADCODE)
/libr/crypto/p/crypto_rc6.c: 33 in rc6_init()
27     
28     	int u = w / 8;
29     	int c = keylen / u;
30     	int t = 2 * r + 4;
31     	ut32 L[c];
32     	ut32 A = 0, B = 0, k = 0, j = 0;
   CID 1355197:  Control flow issues  (DEADCODE)
   Execution cannot reach the expression "c" inside this statement: "v = 3 * ((c > t) ? c : t);".
33     	ut32 v = 3 * ((c > t) ? c : t);
34     
35     	int i, off;
36     
37     	for (i = 0, off = 0; i < c; i++) {
38     		L[i] = ((key[off++] & 0xff));



```

**Comment by Maijin** [comment]
@P4N74 for rc6 and punnycode and @hamishcoleman for arc

**Comment by P4N74** [comment]
fixed `CID 1355198` and `CID 1355199`. will add PR with `CID 1355197` fix too.

**Comment by alvarofe** [comment]
fixed the mach0 ones.

**Comment by hamishcoleman** [comment]
Fixed CID 1355196 and trying to markup CID 1355205 and CID 1355205 so they look OK
